### PR TITLE
Fix possible change of scale sign when decomposing matrix

### PR DESCRIPTION
**Pull request by mackong**
When animate for rotation or scale path, decompose the transform matrix may change the sign of scale, and then compose the matrix back with scale or rotation changed by animation, the matrix may be wrong.

This PR save the origin scale value of node from gltf data, and then use sign of this value as new decomposed scale's sign. If no origin scale vlaue provided, still use determinant's sign.

According to gltf spec, when a node is targeted for animation (referenced by an animation.channel.target), matrix **MUST NOT** be present, so the origin scale value is always available without decompose for a animation node.

Here is a test gltf:
[animation.glb.tar.gz](https://github.com/google/filament/files/12488413/animation.glb.tar.gz)

Before:
image

After:
image

### CODE REVIEW

File: filament/include/filament/TransformManager.h [hunk 1]
```diff
@@ -222,6 +222,23 @@ class UTILS_PUBLIC TransformManager : public FilamentAPI {
...
     children_iterator getChildrenEnd(Instance parent) const noexcept;
 
+    /**
+     * Sets a scale of a transform component.
+     * This scale is preserved for sign check when decomposing transform matrix.
+     * @param ci     The instance of the transform component to set the scale to.
+     * @param scale  The preserved scale.
+     * @see getScale()
+     */
+    void setScale(Instance ci, const math::float3& scale) noexcept;
```
Comment (romainguy) [inline filament/include/filament/TransformManager.h:232]: I'll let @pixelflinger do the full review, but I'm pretty sure we do not want something like this on `TransformManager`.

File: filament/src/components/TransformManager.h [hunk 2]
```diff
@@ -142,6 +149,7 @@ class UTILS_PRIVATE FTransformManager : public TransformManager {
         WORLD,          // world transform
         LOCAL_LO,       // accurate local translation
         WORLD_LO,       // accurate world translation
+        SCALE,          // preserved scaling for sign check when decompose matrix
```
Comment (romainguy) [inline filament/src/components/TransformManager.h:152]: You are storing extra data in the ECS that doesn't seem necessary.

File: libs/gltfio/include/gltfio/math.h [hunk 3]
```diff
@@ -66,8 +67,17 @@ UTILS_PUBLIC inline void decomposeMatrix(const filament::math::mat4f& mat, filam
     float scaley = length(float3({d, e, f}));
     float scalez = length(float3({g, h, i}));
     float3 s = { scalex, scaley, scalez };
-    if (det < 0) {
-        s = -s;
+    if (preservedScale != float3{}) {
+        const float signx = (preservedScale.x < 0) ? -1 : 1;
```
Comment (romainguy) [inline libs/gltfio/include/gltfio/math.h:71]: You are using an entire float3 to store 3 bits of information. All you need is whether each of x/y/z was negative or not.

File: libs/math/include/math/TQuatHelpers.h [hunk 4]
```diff
@@ -244,6 +244,10 @@ class TQuatFunctions {
 
     friend inline
     QUATERNION<T> MATH_PURE slerp(const QUATERNION<T>& p, const QUATERNION<T>& q, T t) {
+        if (p == q) {
```
Comment (romainguy) [inline libs/math/include/math/TQuatHelpers.h:247]: This is unrelated to the rest of this PR

**Review by romainguy — CHANGES_REQUESTED** [review]

File: libs/math/include/math/TQuatHelpers.h (hunk 4 above)
Comment (mackong) [reply libs/math/include/math/TQuatHelpers.h]: Deleted

**Comment by pixelflinger** [comment]
@mackong I understand the problem you're trying to solve, but this is not the correct fix. We do not want to add a partial TRS representation to `TransformManager`. The correct solution is for *you* to write your own "TRS" transform manager, it could for instance store the translation, rotation and scaling of your transform, using the same Entity of your object.
All you'd have to do is use your new TRS manager to calculate the final mat4 and pass that to the transform manager. You could do that just before rendering, once a frame.

**Comment by mackong** [comment]
Oh, i see, i'll try the "TRS" transform manager solution.
